File: packages/engine-community/ios/Sources/EngineCommunity/CommunityEngine.swift
import Foundation

private let engineId = "community@1.0.0"
private let pollIntervalSeconds: TimeInterval = 30
private let confidenceThreshold: Float = 0.5

public final class CommunityEngine {

    private var pollTask: Task<Void, Never>?
    private let bus: ThreatBusProtocol

    public init(bus: ThreatBusProtocol) {
        self.bus = bus
    }

    public func start() {
        guard pollTask == nil else { return }
        pollTask = Task {
            while !Task.isCancelled {
                await runScan()
                try? await Task.sleep(nanoseconds: UInt64(pollIntervalSeconds * 1_000_000_000))
            }
        }
    }

    public func stop() {
        pollTask?.cancel()
        pollTask = nil
    }

    private func runScan() async {
        await withTaskGroup(of: Void.self) { group in
            group.addTask { self.checkJailbreak() }
            group.addTask { self.checkDebugger() }
            group.addTask { self.checkSimulator() }
            group.addTask { self.checkHooks() }
        }
    }

    private func checkJailbreak() {
        let r = JailbreakDetector.detect()
        if r.detected && r.confidence >= confidenceThreshold {
            bus.emit(threatId: "jailbreak", severity: "high", confidence: r.confidence,
                     evidence: r.evidence, engineId: engineId)
        }
    }

    private func checkDebugger() {
        let r = DebuggerDetector.detect()
        if r.detected && r.confidence >= confidenceThreshold {
            bus.emit(threatId: "debugger", severity: "high", confidence: r.confidence,
                     evidence: r.evidence, engineId: engineId)
        }
    }

    private func checkSimulator() {
        let r = SimulatorDetector.detect()
        if r.detected && r.confidence >= confidenceThreshold {
            bus.emit(threatId: "simulator", severity: "medium", confidence: r.confidence,
                     evidence: r.evidence, engineId: engineId)
        }
    }

    private func checkHooks() {
        let r = HookDetector.detect()
        if r.detected && r.confidence >= confidenceThreshold {
            bus.emit(threatId: "hooks", severity: "critical", confidence: r.confidence,
                     evidence: r.evidence, engineId: engineId)
        }
    }
}

public protocol ThreatBusProtocol {
    func emit(threatId: String, severity: String, confidence: Float,
              evidence: [String: String], engineId: String)
}

File: packages/engine-community/android/src/main/kotlin/com/guardian/engine/community/CommunityEngine.kt
package com.guardian.engine.community

import android.content.Context
import com.guardian.rn.ThreatBus
import com.guardian.rn.generated.ThreatId
import com.guardian.rn.generated.Severity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

private const val ENGINE_ID = "community@1.0.0"
private const val POLL_INTERVAL_MS = 30_000L
private const val CONFIDENCE_THRESHOLD = 0.5f

class CommunityEngine(private val context: Context) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var pollJob: Job? = null

    fun start(bus: ThreatBus) {
        if (pollJob?.isActive == true) return
        pollJob = scope.launch {
            while (isActive) {
                runScan(bus)
                delay(POLL_INTERVAL_MS)
            }
        }
    }

    fun stop() {
        pollJob?.cancel()
        pollJob = null
    }

    private fun runScan(bus: ThreatBus) {
        val pm = context.packageManager

        runDetector("root") {
            val r = RootDetector.detect()
            if (r.detected && r.confidence >= CONFIDENCE_THRESHOLD) {
                bus.emit(ThreatId.ROOT, Severity.HIGH, r.confidence, r.evidence, ENGINE_ID)
            }
        }

        runDetector("debugger") {
            val r = DebuggerDetector.detect()
            if (r.detected && r.confidence >= CONFIDENCE_THRESHOLD) {
                bus.emit(ThreatId.DEBUGGER, Severity.HIGH, r.confidence, r.evidence, ENGINE_ID)
            }
        }

        runDetector("emulator") {
            val r = EmulatorDetector.detect()
            if (r.detected && r.confidence >= CONFIDENCE_THRESHOLD) {
                bus.emit(ThreatId.EMULATOR, Severity.MEDIUM, r.confidence, r.evidence, ENGINE_ID)
            }
        }

        runDetector("hooks") {
            val r = HookDetector.detect(pm)
            if (r.detected && r.confidence >= CONFIDENCE_THRESHOLD) {
                bus.emit(ThreatId.HOOKS, Severity.CRITICAL, r.confidence, r.evidence, ENGINE_ID)
            }
        }
    }

    private inline fun runDetector(name: String, block: () -> Unit) {
        try {
            block()
        } catch (e: Exception) {
            // Detector failure is non-fatal; log and continue
        }
    }
}
